File: step_11_heaps/11.2_medium/IB116_n_max_pair_combinations/Solution.java
/**
 * Problem: N Max Pair Combinations
 * Difficulty: MEDIUM | XP: 25
 * Source: InterviewBit
 *
 * Given two arrays A and B of size N, find the N largest possible pair sums
 * A[i] + B[j]. Return them in non-increasing order.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE - Generate all N^2 sums, sort, take top N
    // Time: O(N^2 log N)  |  Space: O(N^2)
    // ============================================================
    public static int[] bruteForce(int[] A, int[] B) {
        int n = A.length;
        int[] sums = new int[n * n];
        int idx = 0;
        for (int a : A) for (int b : B) sums[idx++] = a + b;
        Arrays.sort(sums);
        int[] res = new int[n];
        for (int i = 0; i < n; i++) res[i] = sums[n * n - 1 - i];
        return res;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL - Sort both arrays; max-heap with visited set
    // Time: O(N log N)  |  Space: O(N)
    // Sort A and B descending. Start with (A[0]+B[0], 0, 0) in max-heap.
    // Each pop yields the next largest sum; push (i,j+1) and (i+1,j) if not visited.
    // ============================================================
    public static int[] optimal(int[] A, int[] B) {
        int n = A.length;
        Arrays.sort(A); Arrays.sort(B);
        // Max-heap: [-sum, i, j]
        PriorityQueue<int[]> pq = new PriorityQueue<>((x, y) -> y[0] - x[0]);
        Set<Long> visited = new HashSet<>();

        int si = n - 1, sj = n - 1;
        pq.offer(new int[]{A[si] + B[sj], si, sj});
        visited.add(encode(si, sj, n));

        int[] res = new int[n];
        for (int k = 0; k < n; k++) {
            int[] top = pq.poll();
            int sum = top[0], i = top[1], j = top[2];
            res[k] = sum;
            if (i - 1 >= 0 && visited.add(encode(i - 1, j, n)))
                pq.offer(new int[]{A[i-1] + B[j], i-1, j});
            if (j - 1 >= 0 && visited.add(encode(i, j - 1, n)))
                pq.offer(new int[]{A[i] + B[j-1], i, j-1});
        }
        return res;
    }

    private static long encode(int i, int j, int n) {
        return (long) i * n + j;
    }

    // ============================================================
    // APPROACH 3: BEST - Sort A, for each top-N in A pair with sorted B
    //                    using a single linear pass insight
    // Time: O(N log N)  |  Space: O(N)
    // Sort both. The best N pairs come from the top elements of A x top elements of B.
    // Use heap starting at (n-1, n-1), expand left and down (same as Approach 2
    // but without a visited set by noting we only need one direction per row).
    // ============================================================
    public static int[] best(int[] A, int[] B) {
        // Elegant O(N log N): sort A and B; pair A[i] with B[j] via heap
        // Identical algorithm to optimal but without visited set by tracking
        // only (i, j) with j always = current best for row i
        int n = A.length;
        int[] sortedA = A.clone(), sortedB = B.clone();
        Arrays.sort(sortedA); Arrays.sort(sortedB);

        // heap: [sum, i in A, j in B] — j for each i starts at n-1 and decreases
        PriorityQueue<int[]> pq = new PriorityQueue<>((x, y) -> y[0] - x[0]);
        // Initialize: pair each A[i] with B[n-1] (best B)
        for (int i = 0; i < n; i++) {
            pq.offer(new int[]{sortedA[i] + sortedB[n-1], i, n-1});
        }
        int[] res = new int[n];
        for (int k = 0; k < n; k++) {
            int[] top = pq.poll();
            res[k] = top[0];
            int i = top[1], j = top[2];
            if (j - 1 >= 0) {
                pq.offer(new int[]{sortedA[i] + sortedB[j-1], i, j-1});
            }
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println("=== N Max Pair Combinations ===");

        int[] A = {1, 4, 2, 3};
        int[] B = {2, 5, 1, 6};
        System.out.println("A=" + Arrays.toString(A) + ", B=" + Arrays.toString(B));
        System.out.println("Brute:   " + Arrays.toString(bruteForce(A.clone(), B.clone())));
        System.out.println("Optimal: " + Arrays.toString(optimal(A.clone(), B.clone())));
        System.out.println("Best:    " + Arrays.toString(best(A.clone(), B.clone())));
        // Expected: top 4 sums from {1,2,3,4}x{1,2,5,6} = 4+6=10, 4+5=9, 3+6=9, 4+2=6 -> [10,9,9,8]

        int[] A2 = {3, 2};
        int[] B2 = {1, 4};
        System.out.println("\nA=" + Arrays.toString(A2) + ", B=" + Arrays.toString(B2));
        System.out.println("Brute:   " + Arrays.toString(bruteForce(A2.clone(), B2.clone())));
        System.out.println("Optimal: " + Arrays.toString(optimal(A2.clone(), B2.clone())));
        System.out.println("Best:    " + Arrays.toString(best(A2.clone(), B2.clone())));
        // Expected: [7, 6]
    }
}

File: step_11_heaps/11.2_medium/IB116_n_max_pair_combinations/solution.py
"""
Problem: N Max Pair Combinations
Difficulty: MEDIUM | XP: 25
Source: InterviewBit

Given two arrays A and B of size N, find the N largest possible pair sums
A[i] + B[j]. Return them in non-increasing order.
"""
from typing import List
import heapq


# ============================================================
# APPROACH 1: BRUTE FORCE - Generate all N^2 sums, sort, take top N
# Time: O(N^2 log N)  |  Space: O(N^2)
# ============================================================
def brute_force(A: List[int], B: List[int]) -> List[int]:
    n = len(A)
    all_sums = sorted([a + b for a in A for b in B], reverse=True)
    return all_sums[:n]


# ============================================================
# APPROACH 2: OPTIMAL - Sort + max-heap with visited set
# Time: O(N log N)  |  Space: O(N)
# Sort both desc. Start heap with (A[0]+B[0], 0, 0).
# Each extraction gives next largest; push neighbors (i+1,j) and (i,j+1).
# ============================================================
def optimal(A: List[int], B: List[int]) -> List[int]:
    n = len(A)
    A.sort(reverse=True)
    B.sort(reverse=True)

    heap = [-(A[0] + B[0]), 0, 0]  # start with largest pair
    heap = [(-(A[0] + B[0]), 0, 0)]
    visited = {(0, 0)}
    res = []
    while len(res) < n:
        neg_sum, i, j = heapq.heappop(heap)
        res.append(-neg_sum)
        if i + 1 < n and (i + 1, j) not in visited:
            visited.add((i + 1, j))
            heapq.heappush(heap, (-(A[i+1] + B[j]), i + 1, j))
        if j + 1 < n and (i, j + 1) not in visited:
            visited.add((i, j + 1))
            heapq.heappush(heap, (-(A[i] + B[j+1]), i, j + 1))
    return res


# ============================================================
# APPROACH 3: BEST - Sort A & B; initialize heap with each A[i] paired with B[0]
# Time: O(N log N)  |  Space: O(N)
# No visited set needed: each A[i] independently walks down sorted B
# ============================================================
def best(A: List[int], B: List[int]) -> List[int]:
    n = len(A)
    A_sorted = sorted(A, reverse=True)
    B_sorted = sorted(B, reverse=True)

    # Each row i: A_sorted[i] starts paired with B_sorted[0]
    heap = [-(A_sorted[i] + B_sorted[0]) for i in range(n)]
    # Also track which B index each A index is at
    heap_full = [(-( A_sorted[i] + B_sorted[0]), i, 0) for i in range(n)]
    heapq.heapify(heap_full)

    res = []
    while len(res) < n:
        neg_sum, ai, bi = heapq.heappop(heap_full)
        res.append(-neg_sum)
        if bi + 1 < n:
            heapq.heappush(heap_full, (-(A_sorted[ai] + B_sorted[bi + 1]), ai, bi + 1))
    return res


if __name__ == "__main__":
    print("=== N Max Pair Combinations ===")

    A, B = [1, 4, 2, 3], [2, 5, 1, 6]
    print(f"A={A}, B={B}")
    print(f"Brute:   {brute_force(A[:], B[:])}")
    print(f"Optimal: {optimal(A[:], B[:])}")
    print(f"Best:    {best(A[:], B[:])}")
    # Expected: top 4 = [10, 9, 9, 8] (4+6, 3+6 or 4+5, ...)

    A2, B2 = [3, 2], [1, 4]
    print(f"\nA={A2}, B={B2}")
    print(f"Brute:   {brute_force(A2[:], B2[:])}")
    print(f"Optimal: {optimal(A2[:], B2[:])}")
    print(f"Best:    {best(A2[:], B2[:])}")
    # Expected: [7, 6]
